src/Plugin/Field/FieldType/EntityReferenceOverrideItem.php

```diff
@@ -0,0 +1,167 @@
+<?php
+
+namespace Drupal\entity_reference_override\Plugin\Field\FieldType;
+
+use Drupal\Component\Utility\NestedArray;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Field\FieldStorageDefinitionInterface;
+use Drupal\Core\Field\Plugin\Field\FieldType\EntityReferenceItem;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\TypedData\MapDataDefinition;
+
+/**
+ * Plugin implementation of the 'media' field type.
+ *
+ * @FieldType(
+ *   id = "entity_reference_override",
+ *   label = @Translation("Entity reference override"),
+ *   description = @Translation("An entity field containing an entity reference and additional data."),
+ *   category = @Translation("Reference"),
+ *   default_widget = "entity_reference_autocomplete",
+ *   default_formatter = "entity_reference_label",
+ *   list_class = "\Drupal\Core\Field\EntityReferenceFieldItemList"
+ * )
+ */
+class EntityReferenceOverrideItem extends EntityReferenceItem {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function propertyDefinitions(FieldStorageDefinitionInterface $field_definition) {
+    $properties = parent::propertyDefinitions($field_definition);
+
+    $properties['overwritten_property_map'] = MapDataDefinition::create()
+      ->setLabel(t('Overwritten property map'));
+
+    return $properties;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function schema(FieldStorageDefinitionInterface $field_definition) {
+    $schema = parent::schema($field_definition);
+
+    $schema['columns']['overwritten_property_map'] = [
+      'description' => 'A map to overwrite media data per instance.',
+      'type' => 'blob',
+      'size' => 'big',
+      'serialize' => TRUE,
+    ];
+
+    return $schema;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __get($name) {
+    if ($name == 'entity' && !empty($this->values['overwritten_property_map'])) {
+      /** @var \Drupal\Core\Entity\ContentEntityInterface $entity */
+      $entity = parent::__get('entity');
+
+      $this->overwriteFields($entity, $this->values['overwritten_property_map']);
+      $translation = $entity->getTranslation($this->getLangcode());
+      $this->overwriteFields($translation, $this->values['overwritten_property_map']);
+
+      $entity->addCacheableDependency($this->getEntity());
+      $entity->overwritten = TRUE;
+      return $entity;
+    }
+    return parent::__get($name);
+  }
+
+  /**
+   * Override entity fields.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The entity to override.
+   * @param array $overwritten_property_map
+   *   The new values.
+   */
+  protected function overwriteFields(EntityInterface $entity, array $overwritten_property_map) {
+    foreach ($overwritten_property_map as $key => $value) {
+      if (is_array($value)) {
+        // Remove keys that don't exists in original media entity.
+        $values = array_intersect_key($value, $entity->get($key)->getValue());
```

Comment:
I wouldn't have $value and $values - it is a bit confusing - I'd reassign $value here and then you can remove the else below.